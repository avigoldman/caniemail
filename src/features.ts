import { getProperty } from 'dot-prop';
import onetime from 'onetime';

import { type EmailClient, type SupportTypeResult, getSupportType } from './clients.js';
import { type RawFeatureData, caniEmailJson } from './json.cjs';

export { caniEmailJson as rawData };

interface PositionPoint {
  line: number;
  column: number;
}

export interface Position {
  start: PositionPoint;
  end: PositionPoint;
  source?: string;
}

export interface FeatureIssue {
  notes: string[];
  title: string;
  position?: Position;
  support: SupportTypeResult['type'];
}

export interface FeatureInfo extends FeatureIssue {
  url: string;
}

export interface FeatureIssues {
  errors: FeatureMap<FeatureIssue>;
  warnings: FeatureMap<FeatureIssue>;
}

type FeatureSet = Map<string, RawFeatureData>;

interface GetFeaturesResult {
  all: FeatureSet;
  css: FeatureSet;
  html: FeatureSet;
}

export class FeatureMap<TValue> extends Map<EmailClient, TValue[]> {
  // @ts-ignore
  set(key: EmailClient, value: TValue): this {
    const values = super.get(key) || [];
    values.push(value);
    super.set(key, values);
    return this;
  }
}

export const getFeatures = onetime((): GetFeaturesResult => {
  const all = new Map();
  const css = new Map();
  const html = new Map();

  for (const bit of caniEmailJson.data) {
    all.set(bit.title, bit);
    if (bit.category === 'css') css.set(bit.title, bit);
    if (bit.category === 'html') html.set(bit.title, bit);
  }

  return {
    all,
    css,
    html
  };
});

export const getAllFeatures = (clients: EmailClient[]) => {
  const results = {
    supported: new FeatureMap<FeatureInfo>(),
    unsupported: new FeatureMap<FeatureInfo>()
  };
  const { all: features } = getFeatures();

  for (const [, feature] of features) {
    const { stats, title, url } = feature;

    for (const client of clients) {
      const supportMap = getProperty(stats, client);

      // eslint-disable-next-line no-continue
      if (supportMap === void 0) continue;

      const supportStatus = getSupportType(supportMap);
      const notes = (supportStatus.noteNumbers ?? []).map(
        (noteNumber: number) => feature.notes_by_num![String(noteNumber)]
      );
      const support = supportStatus.type;

      if (support === 'none') {
        results.unsupported.set(client, { notes, support, title, url });
      } else {
        results.supported.set(client, { notes, support, title, url });
      }
    }
  }

  return results;
};
