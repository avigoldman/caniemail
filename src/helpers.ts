import css, { type CssStylesheetAST } from '@adobe/css-tools';
import type { Document, Element, Node, Text } from 'domhandler';
import * as htmlparser from 'htmlparser2';

interface ParseHtmlResult {
  document: Document;
  stylesheets: CssStylesheetAST[];
}

export function fromTitleEntries<T>(
  entries: Array<{ title: string; value: T } | undefined>
): Record<string, T> {
  return Object.fromEntries(
    entries.filter((entry) => entry !== void 0).map((entry) => [entry.title, entry.value])
  );
}

export const getTitleMatches = (
  pairs: Record<string, any>,
  targetValue: string | string[],
  comparer?: (value: any, targetValue: string | string[]) => Boolean
): string[] =>
  Object.entries(pairs)
    .filter(([, value]) => (comparer ? comparer(value, targetValue) : targetValue.includes(value)))
    .map(([title]) => title);

const getStyleNodes = (node: Node) => {
  if (node.type === htmlparser.ElementType.Style) return [node as Element];

  const results: Element[] = [];

  if ('childNodes' in node) {
    for (const child of (node as Element).childNodes) {
      results.push(...getStyleNodes(child));
    }
  }

  return results;
};

export function parseHtml(html: string): ParseHtmlResult {
  const document = htmlparser.parseDocument(html);

  const styleNodes: Element[] = [];
  for (const childNode of document.childNodes) {
    styleNodes.push(...getStyleNodes(childNode));
  }

  const stylesheets: CssStylesheetAST[] = [];
  for (const styleNode of styleNodes) {
    const styleTextNode = styleNode.childNodes[0] as Text | undefined;
    if (styleTextNode !== void 0) {
      stylesheets.push(css.parse(styleTextNode.data));
    }
  }

  return {
    document,
    stylesheets
  };
}
