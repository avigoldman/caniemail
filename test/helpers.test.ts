import { describe, expect, test } from 'vitest';
import { FeatureMap, type FeatureIssue } from '../dist/features';
import { groupIssues, sortIssues, type IssueGroup } from '../dist/helpers';

describe('issue formatting', () => {
  test('groups single map of issues correctly', () => {
    const errors = new FeatureMap<FeatureIssue>();

    errors.set('outlook.windows', {
      title: 'Parse Error',
      position: { start: { line: 1, column: 10 }, end: { line: 1, column: 15 } },
      notes: ['Invalid syntax'],
      support: 'none'
    });
    errors.set('gmail.ios', {
      title: 'Parse Error',
      position: { start: { line: 1, column: 10 }, end: { line: 1, column: 15 } },
      notes: ['Invalid syntax'],
      support: 'none'
    });
    errors.set('yahoo.ios', {
      title: 'Parse Error',
      position: { start: { line: 2, column: 5 }, end: { line: 2, column: 10 } },
      notes: ['Different position, same title'],
      support: 'none'
    });

    const grouped = groupIssues(errors);

    expect(grouped).toHaveLength(2);
    expect(grouped).toMatchSnapshot();
  });

  test('sorts grouped issues correctly', () => {
    const unsortedIssues: IssueGroup[] = [
      // Line 5 issues
      {
        issue: {
          title: 'Late Error',
          position: { start: { line: 5, column: 15 }, end: { line: 5, column: 20 } },
          notes: [],
          support: 'none'
        },
        clients: ['yahoo.ios']
      },
      {
        issue: {
          title: 'Mid-Line Warning',
          position: { start: { line: 5, column: 8 }, end: { line: 5, column: 12 } },
          notes: [],
          support: 'partial'
        },
        clients: ['outlook.ios']
      },
      {
        issue: {
          title: 'Early Error',
          position: { start: { line: 5, column: 1 }, end: { line: 5, column: 5 } },
          notes: [],
          support: 'none'
        },
        clients: ['gmail.ios']
      },
      // Line 3 issues
      {
        issue: {
          title: 'Late Warning',
          position: { start: { line: 3, column: 20 }, end: { line: 3, column: 25 } },
          notes: [],
          support: 'partial'
        },
        clients: ['thunderbird.macos']
      },
      {
        issue: {
          title: 'Early Line 3',
          position: { start: { line: 3, column: 5 }, end: { line: 3, column: 10 } },
          notes: [],
          support: 'none'
        },
        clients: ['protonmail.ios']
      },
      // Line 8 issues
      {
        issue: {
          title: 'Very Late Warning',
          position: { start: { line: 8, column: 12 }, end: { line: 8, column: 15 } },
          notes: [],
          support: 'partial'
        },
        clients: ['apple-mail.ios']
      },
      // Line 1 issues
      {
        issue: {
          title: 'First Error',
          position: { start: { line: 1, column: 10 }, end: { line: 1, column: 15 } },
          notes: [],
          support: 'none'
        },
        clients: ['outlook.ios']
      },
      {
        issue: {
          title: 'First Warning',
          position: { start: { line: 1, column: 5 }, end: { line: 1, column: 8 } },
          notes: [],
          support: 'partial'
        },
        clients: ['gmail.ios']
      },
      // Line 7 issue
      {
        issue: {
          title: 'Lone Error',
          position: { start: { line: 7, column: 1 }, end: { line: 7, column: 5 } },
          notes: [],
          support: 'none'
        },
        clients: ['yahoo.ios']
      }
    ];

    const sorted = sortIssues(unsortedIssues);

    expect(
      sorted.map(
        ({ issue }) =>
          `${issue.support} ${issue.position!.start.line}:${issue.position!.start.column}`
      )
    ).toEqual([
      'partial 1:5',
      'none 1:10',
      'none 3:5',
      'partial 3:20',
      'none 5:1',
      'partial 5:8',
      'none 5:15',
      'none 7:1',
      'partial 8:12'
    ]);
  });
});
