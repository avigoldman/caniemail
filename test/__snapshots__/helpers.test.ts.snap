// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`issue formatting > groups single map of issues correctly 1`] = `
[
  {
    "clients": [
      "outlook.windows",
      "gmail.ios",
    ],
    "issue": {
      "notes": [
        "Invalid syntax",
      ],
      "position": {
        "end": {
          "column": 15,
          "line": 1,
        },
        "start": {
          "column": 10,
          "line": 1,
        },
      },
      "support": "none",
      "title": "Parse Error",
    },
  },
  {
    "clients": [
      "yahoo.ios",
    ],
    "issue": {
      "notes": [
        "Different position, same title",
      ],
      "position": {
        "end": {
          "column": 10,
          "line": 2,
        },
        "start": {
          "column": 5,
          "line": 2,
        },
      },
      "support": "none",
      "title": "Parse Error",
    },
  },
]
`;
